src/gar.lua

```diff
@@ -1191,4 +1196,259 @@ function gar.delegateAllowedToStake(delegateAddress, gateway)
 		or (gateway.delegates[delegateAddress] and gateway.delegates[delegateAddress].delegatedStake or 0) > 0
 end
 
+function gar.getFundingPlan(address, quantity, sourcesPreference)
+	sourcesPreference = sourcesPreference or "balance"
+	local fundingPlan = {
+		address = address,
+		balance = 0,
+		stakes = {},
+		shortfall = quantity,
+	}
+
+	planBalanceDrawdown(fundingPlan, sourcesPreference)
+
+	-- if the remaining quantity is 0 or there are no more sources, return early
+	if fundingPlan.shortfall == 0 or sourcesPreference == "balance" then
+		return fundingPlan
+	end
+
+	local gatewaysInfo = planExcessStakesDrawdown(fundingPlan)
+
+	-- early return if possible. Otherwise we'll move on to use delegation vaults
+	if fundingPlan.shortfall == 0 then
+		return fundingPlan
+	end
+
+	planVaultsDrawdown(fundingPlan, gatewaysInfo)
+
+	-- early return if possible. Otherwise we'll move on to using minimum stakes
+	if fundingPlan.shortfall == 0 then
+		return fundingPlan
+	end
+
+	planMinimumStakesDrawdown(fundingPlan, gatewaysInfo)
+
+	return fundingPlan
+end
+
+function planBalanceDrawdown(fundingPlan, sourcesPreference)
+	local availableBalance = balances.getBalance(fundingPlan.address)
+	if sourcesPreference == "balance" or sourcesPreference == "any" then
+		fundingPlan.balance = math.min(availableBalance, fundingPlan.shortfall)
+		fundingPlan.shortfall = fundingPlan.shortfall - fundingPlan.balance
+	end
+end
```

Comment:
note: bc these are not exported you likely cannot unit test

Reply:
Correct - the are tested in integration with the top level function.